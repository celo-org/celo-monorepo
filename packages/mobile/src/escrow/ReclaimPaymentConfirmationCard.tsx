import HorizontalLine from '@celo/react-components/components/HorizontalLine'
import { MoneyAmount } from '@celo/react-components/components/MoneyAmount'
import PhoneNumberWithFlag from '@celo/react-components/components/PhoneNumberWithFlag'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import { componentStyles } from '@celo/react-components/styles/styles'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { withNamespaces, WithNamespaces } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { connect } from 'react-redux'
import LineItemRow from 'src/components/LineItemRow'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import { Namespaces } from 'src/i18n'
import Logo from 'src/icons/Logo'
import { RecipientWithContact } from 'src/recipients/recipient'
import { RootState } from 'src/redux/reducers'
import FeeIcon from 'src/send/FeeIcon'
import { getFeeDisplayValue, getMoneyDisplayValue } from 'src/utils/formatting'

export interface OwnProps {
  recipientPhone: string
  recipientContact?: RecipientWithContact
  amount: BigNumber
  fee?: BigNumber
  isLoadingFee?: boolean
  feeError?: Error
  currency: CURRENCY_ENUM
}

interface StateProps {
  defaultCountryCode: string
}

const mapStateToProps = (state: RootState): StateProps => {
  return {
    defaultCountryCode: state.account.defaultCountryCode,
  }
}

type Props = OwnProps & StateProps & WithNamespaces

class ReclaimPaymentConfirmationCard extends React.PureComponent<Props> {
  render() {
    const {
      recipientPhone,
      recipientContact,
      amount,
      fee,
      isLoadingFee,
      feeError,
      currency,
      defaultCountryCode,
      t,
    } = this.props
    const currencySymbol = CURRENCIES[currency].symbol
    const amountWithFees = getMoneyDisplayValue(amount.minus(fee || 0))

    return (
      <View style={[componentStyles.roundedBorder, style.container]}>
        <View style={style.logo}>
          <Logo height={40} />
        </View>
        <MoneyAmount
          symbol={currencySymbol}
          amount={amountWithFees}
          sign={'+'}
          color={colors.celoGreen}
        />
        <HorizontalLine />
        <View style={style.details}>
          {recipientContact && (
            <Text style={[fontStyles.bodySmallSemiBold, style.contactName]}>
              {recipientContact.displayName}
            </Text>
          )}
          {recipientPhone && (
            <PhoneNumberWithFlag
              e164PhoneNumber={recipientPhone}
              defaultCountryCode={defaultCountryCode}
            />
          )}
        </View>
        <View style={style.feeContainer}>
          <LineItemRow
            currencySymbol={currencySymbol}
            amount={amount.toString()}
            title={t('totalSent')}
          />
          <LineItemRow
            currencySymbol={currencySymbol}
            amount={getFeeDisplayValue(fee)}
            title={t('securityFee')}
            titleIcon={<FeeIcon />}
            isLoading={isLoadingFee}
            hasError={!!feeError}
          />
          <LineItemRow
            currencySymbol={currencySymbol}
            amount={amountWithFees}
            title={t('totalRefunded')}
          />
        </View>
      </View>
    )
  }
}

const style = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 20,
  },
  feeContainer: {
    marginBottom: 10,
    justifyContent: 'center',
    alignItems: 'stretch',
  },
  contactName: {
    paddingTop: 6,
    textAlign: 'center',
  },
  logo: {
    alignSelf: 'center',
    margin: 'auto',
    marginVertical: 10,
  },
  details: {
    padding: 20,
  },
})

export default connect<StateProps, {}, {}, RootState>(mapStateToProps)(
  withNamespaces(Namespaces.sendFlow7)(ReclaimPaymentConfirmationCard)
)
