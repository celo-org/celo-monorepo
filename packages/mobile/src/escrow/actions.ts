import BigNumber from 'bignumber.js'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { SHORT_CURRENCIES } from 'src/geth/consts'
import { RecipientWithContact } from 'src/recipients/recipient'

export interface EscrowedPayment {
  senderAddress: string
  recipientPhone: string
  recipientContact?: RecipientWithContact
  paymentID: string
  currency: SHORT_CURRENCIES
  amount: BigNumber
  message?: string
  timestamp: BigNumber
  expirySeconds: BigNumber
}

export enum Actions {
  TRANSFER_PAYMENT = 'ESCROW/TRANSFER_PAYMENT',
  RECLAIM_PAYMENT = 'ESCROW/RECLAIM_PAYMENT',
  FETCH_SENT_PAYMENTS = 'ESCROW/FETCH_SENT_PAYMENTS',
  STORE_SENT_PAYMENTS = 'ESCROW/STORE_SENT_PAYMENTS',
  RESEND_PAYMENT = 'ESCROW/RESEND_PAYMENT',
  RECLAIM_PAYMENT_SUCCESS = 'ESCROW/RECLAIM_PAYMENT_SUCCESS',
  RECLAIM_PAYMENT_FAILURE = 'ESCROW/RECLAIM_PAYMENT_FAILURE',
}

export interface EscrowTransferPaymentAction {
  type: Actions.TRANSFER_PAYMENT
  phoneHash: string
  amount: BigNumber
  tempWalletAddress: string
}
export interface EscrowReclaimPaymentAction {
  type: Actions.RECLAIM_PAYMENT
  paymentID: string
}

export interface EscrowFetchSentPaymentsAction {
  type: Actions.FETCH_SENT_PAYMENTS
  forceRefresh: boolean
}

export interface EscrowStoreSentPaymentsAction {
  type: Actions.STORE_SENT_PAYMENTS
  sentPayments: EscrowedPayment[]
}

export interface EscrowResendPaymentAction {
  type: Actions.RESEND_PAYMENT
  paymentId: string
}

export interface EscrowReclaimPaymentSuccessAction {
  type: Actions.RECLAIM_PAYMENT_SUCCESS
}

export interface EscrowReclaimFailureAction {
  type: Actions.RECLAIM_PAYMENT_FAILURE
  error: ErrorMessages
}

export type ActionTypes =
  | EscrowTransferPaymentAction
  | EscrowReclaimPaymentAction
  | EscrowFetchSentPaymentsAction
  | EscrowStoreSentPaymentsAction
  | EscrowResendPaymentAction
  | EscrowReclaimPaymentSuccessAction
  | EscrowReclaimFailureAction

export const transferEscrowedPayment = (
  phoneHash: string,
  amount: BigNumber,
  tempWalletAddress: string
): EscrowTransferPaymentAction => ({
  type: Actions.TRANSFER_PAYMENT,
  phoneHash,
  amount,
  tempWalletAddress,
})

export const reclaimEscrowPayment = (paymentID: string): EscrowReclaimPaymentAction => ({
  type: Actions.RECLAIM_PAYMENT,
  paymentID,
})

export const fetchSentEscrowPayments = (forceRefresh = false): EscrowFetchSentPaymentsAction => ({
  type: Actions.FETCH_SENT_PAYMENTS,
  forceRefresh,
})

export const storeSentEscrowPayments = (
  sentPayments: EscrowedPayment[]
): EscrowStoreSentPaymentsAction => ({
  type: Actions.STORE_SENT_PAYMENTS,
  sentPayments,
})

export const resendEscrowPayment = (paymentId: string): EscrowResendPaymentAction => ({
  type: Actions.RESEND_PAYMENT,
  paymentId,
})

export const reclaimEscrowPaymentSuccess = (): EscrowReclaimPaymentSuccessAction => ({
  type: Actions.RECLAIM_PAYMENT_SUCCESS,
})

export const reclaimEscrowPaymentFailure = (error: ErrorMessages): EscrowReclaimFailureAction => ({
  type: Actions.RECLAIM_PAYMENT_FAILURE,
  error,
})
