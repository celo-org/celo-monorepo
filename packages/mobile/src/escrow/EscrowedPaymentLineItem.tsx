import fontStyles from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { StyleSheet, Text } from 'react-native'
import { EscrowedPayment } from 'src/escrow/actions'
import { divideByWei, getCentAwareMoneyDisplay } from 'src/utils/formatting'

interface Props {
  payment: EscrowedPayment
}

export default function EscrowedPaymentLineItem(props: Props) {
  const { amount, message, recipientPhone } = props.payment
  return (
    <Text numberOfLines={1} ellipsizeMode="middle" style={styles.oneLine}>
      <Text style={[fontStyles.subSmall]}>
        {recipientPhone} - {message}
      </Text>
      <Text style={[fontStyles.subSmall, fontStyles.semiBold]}>
        {' '}
        ${getCentAwareMoneyDisplay(divideByWei(amount.toString()))}
      </Text>
    </Text>
  )
}

const styles = StyleSheet.create({
  oneLine: {
    flexDirection: 'row',
  },
})
