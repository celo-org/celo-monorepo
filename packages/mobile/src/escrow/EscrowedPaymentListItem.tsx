import BaseNotification from '@celo/react-components/components/BaseNotification'
import fontStyles from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Image, Platform, StyleSheet, Text, View } from 'react-native'
import SendIntentAndroid from 'react-native-send-intent'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { componentWithAnalytics } from 'src/analytics/wrapper'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { EscrowedPayment } from 'src/escrow/actions'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import { Namespaces } from 'src/i18n'
import { inviteFriendsIcon } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { divideByWei, getCentAwareMoneyDisplay } from 'src/utils/formatting'
import { navigateToURI } from 'src/utils/linking'
import Logger from 'src/utils/Logger'

interface OwnProps {
  payment: EscrowedPayment
}

type Props = OwnProps & WithNamespaces

const TAG = 'EscrowedPaymentListItem'

export class EscrowedPaymentListItem extends React.PureComponent<Props> {
  onRemind = () => {
    const { payment, t } = this.props
    const recipientPhoneNumber = payment.recipientPhone
    CeloAnalytics.track(CustomEventNames.clicked_escrowed_payment_send_message)
    // TODO: open up whatsapp/text message slider with pre populated message
    try {
      if (Platform.OS === 'android') {
        SendIntentAndroid.sendSms(recipientPhoneNumber, t('walletFlow5:escrowedPaymentReminderSms'))
      } else {
        // TODO look into using MFMessageComposeViewController to prefill the body for iOS
        navigateToURI(`sms:${recipientPhoneNumber}`)
      }
    } catch (error) {
      // TODO: use the showError saga instead of the Logger.showError, which is a hacky temp thing we used for a while that doesn't actually work on iOS
      Logger.showError(ErrorMessages.SMS_ERROR)
      Logger.error(TAG, `Error sending SMS to ${recipientPhoneNumber}`, error)
    }
  }
  onReclaimPayment = () => {
    const { payment } = this.props
    const reclaimPaymentInput = payment
    CeloAnalytics.track(CustomEventNames.clicked_escrowed_payment_notification)
    navigate(Screens.ReclaimPaymentConfirmationScreen, { reclaimPaymentInput })
  }
  getCTA = () => {
    const { t } = this.props
    const ctas = []
    if (this.getDisplayName()) {
      ctas.push({
        text: t('global:remind'),
        onPress: this.onRemind,
      })
    }
    ctas.push({
      text: t('global:reclaim'),
      onPress: this.onReclaimPayment,
    })
    return ctas
  }

  getDisplayName() {
    const { payment } = this.props
    return payment.recipientContact ? payment.recipientContact.displayName : payment.recipientPhone
  }

  render() {
    const { t, payment } = this.props
    return (
      <View style={styles.container}>
        <BaseNotification
          title={t('escrowPaymentNotificationTitle', {
            mobile: this.getDisplayName() || t('global:unknown'),
            amount:
              CURRENCIES[CURRENCY_ENUM.DOLLAR].symbol +
              getCentAwareMoneyDisplay(divideByWei(payment.amount)),
          })}
          icon={<Image source={inviteFriendsIcon} style={styles.image} resizeMode="contain" />}
          ctas={this.getCTA()}
        >
          <Text style={fontStyles.bodySmall}>{payment.message || t('defaultComment')}</Text>
        </BaseNotification>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
  body: {
    marginTop: 5,
    flexDirection: 'row',
  },
  image: {
    width: 30,
    height: 30,
  },
  payment: {
    flex: 1,
  },
})

export default componentWithAnalytics(
  withNamespaces(Namespaces.inviteFlow11)(EscrowedPaymentListItem)
)
