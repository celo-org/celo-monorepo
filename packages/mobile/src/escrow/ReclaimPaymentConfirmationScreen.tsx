import ReviewFrame from '@celo/react-components/components/ReviewFrame'
import ReviewHeader from '@celo/react-components/components/ReviewHeader'
import colors from '@celo/react-components/styles/colors'
import { CURRENCY_ENUM } from '@celo/utils/src/currencies'
import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { ActivityIndicator, StyleSheet } from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import { NavigationInjectedProps } from 'react-navigation'
import { connect } from 'react-redux'
import { showError } from 'src/alert/actions'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import componentWithAnalytics from 'src/analytics/wrapper'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { EscrowedPayment, reclaimEscrowPayment } from 'src/escrow/actions'
import ReclaimPaymentConfirmationCard from 'src/escrow/ReclaimPaymentConfirmationCard'
import { FeeType } from 'src/fees/actions'
import CalculateFee, { CalculateFeeChildren } from 'src/fees/CalculateFee'
import { getFeeDollars } from 'src/fees/selectors'
import { Namespaces, withTranslation } from 'src/i18n'
import { navigateBack } from 'src/navigator/NavigationService'
import { RootState } from 'src/redux/reducers'
import { isAppConnected } from 'src/redux/selectors'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import { divideByWei } from 'src/utils/formatting'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'escrow/ReclaimPaymentConfirmationScreen'

interface StateProps {
  isReclaiming: boolean
  e164PhoneNumber: string
  account: string | null
  dollarBalance: string
  appConnected: boolean
}

interface DispatchProps {
  reclaimPayment: typeof reclaimEscrowPayment
  showError: typeof showError
}

const mapDispatchToProps = {
  reclaimPayment: reclaimEscrowPayment,
  showError,
}

const mapStateToProps = (state: RootState): StateProps => {
  return {
    isReclaiming: state.escrow.isReclaiming,
    e164PhoneNumber: state.account.e164PhoneNumber,
    account: currentAccountSelector(state),
    dollarBalance: state.stableToken.balance || '0',
    appConnected: isAppConnected(state),
  }
}

type Props = NavigationInjectedProps & DispatchProps & StateProps & WithTranslation

class ReclaimPaymentConfirmationScreen extends React.Component<Props> {
  static navigationOptions = { header: null }

  getReclaimPaymentInput(): EscrowedPayment {
    const reclaimPaymentInput = this.props.navigation.getParam('reclaimPaymentInput', '')
    if (reclaimPaymentInput === '') {
      throw new Error('Reclaim payment input missing')
    }
    return reclaimPaymentInput
  }

  onConfirm = async () => {
    const escrowedPayment = this.getReclaimPaymentInput()
    CeloAnalytics.track(CustomEventNames.escrowed_payment_reclaimed_by_sender)
    const address = this.props.account
    if (!address) {
      throw new Error("Can't reclaim funds without a valid account")
    }

    try {
      this.props.reclaimPayment(escrowedPayment.paymentID)
    } catch (error) {
      Logger.error(TAG, 'Reclaiming escrowed payment failed, show error message', error)
      this.props.showError(ErrorMessages.RECLAIMING_ESCROWED_PAYMENT_FAILED)
      return
    }
  }

  onPressEdit = () => {
    CeloAnalytics.track(CustomEventNames.escrowed_payment_reclaimEdit_by_sender)
    navigateBack()
  }

  renderHeader = () => {
    const { t } = this.props
    const title = t('reclaimPayment')
    return <ReviewHeader title={title} />
  }

  renderFooter = () => {
    return this.props.isReclaiming ? (
      <ActivityIndicator size="large" color={colors.celoGreen} />
    ) : null
  }

  renderWithAsyncFee: CalculateFeeChildren = (asyncFee) => {
    const { t, isReclaiming, appConnected, dollarBalance } = this.props
    const payment = this.getReclaimPaymentInput()
    const fee = getFeeDollars(asyncFee.result)
    const convertedAmount = divideByWei(payment.amount.valueOf())
    const userHasEnough = fee && fee.isLessThanOrEqualTo(dollarBalance)

    return (
      <SafeAreaView style={styles.container}>
        <DisconnectBanner />
        <ReviewFrame
          HeaderComponent={this.renderHeader}
          FooterComponent={this.renderFooter}
          confirmButton={{
            action: this.onConfirm,
            text: t('global:confirm'),
            disabled:
              isReclaiming ||
              !userHasEnough ||
              !appConnected ||
              asyncFee.loading ||
              !!asyncFee.error,
          }}
          modifyButton={{ action: this.onPressEdit, text: t('cancel'), disabled: isReclaiming }}
        >
          <ReclaimPaymentConfirmationCard
            recipientPhone={payment.recipientPhone}
            recipientContact={
              undefined /* TODO get recipient contact details from recipient cache*/
            }
            amount={convertedAmount}
            currency={CURRENCY_ENUM.DOLLAR} // User can only request in Dollars
            fee={fee}
            isLoadingFee={asyncFee.loading}
            feeError={asyncFee.error}
          />
        </ReviewFrame>
      </SafeAreaView>
    )
  }

  render() {
    const { account } = this.props
    if (!account) {
      throw Error('Account is required')
    }

    const payment = this.getReclaimPaymentInput()

    return (
      // Note: intentionally passing a new child func here otherwise
      // it doesn't re-render on state change since CalculateFee is a pure component
      <CalculateFee
        feeType={FeeType.RECLAIM_ESCROW}
        account={account}
        paymentID={payment.paymentID}
      >
        {(asyncFee) => this.renderWithAsyncFee(asyncFee)}
      </CalculateFee>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
    paddingTop: 20,
  },
})

export default componentWithAnalytics(
  connect<StateProps, DispatchProps, {}, RootState>(
    mapStateToProps,
    mapDispatchToProps
  )(withTranslation(Namespaces.sendFlow7)(ReclaimPaymentConfirmationScreen))
)
