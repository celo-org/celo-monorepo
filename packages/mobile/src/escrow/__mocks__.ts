import BigNumber from 'bignumber.js'
import { EscrowedPayment } from 'src/escrow/actions'
import { SHORT_CURRENCIES } from 'src/geth/consts'
import { multiplyByWei } from 'src/utils/formatting'
import { mockInvitableRecipient } from 'test/values'

const recipientPhone = '+491522345678'
const senderAddress = '0x000000000000000000000ce10'
const currency = SHORT_CURRENCIES.DOLLAR

const date = new BigNumber(
  new Date('Tue Mar 05 2019 13:44:06 GMT-0800 (Pacific Standard Time)').getTime()
)

export function escrowPaymentDouble(partial: object): EscrowedPayment {
  return {
    senderAddress,
    recipientPhone,
    recipientContact: mockInvitableRecipient,
    paymentID: 'FAKE_ID_1',
    currency,
    amount: multiplyByWei(new BigNumber(7)),
    message: 'test message',
    timestamp: date,
    expirySeconds: new BigNumber(60),
    ...partial,
  }
}
