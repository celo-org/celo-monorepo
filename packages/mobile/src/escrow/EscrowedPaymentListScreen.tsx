import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { View } from 'react-native'
import { NavigationInjectedProps } from 'react-navigation'
import { connect } from 'react-redux'
import { EscrowedPayment } from 'src/escrow/actions'
import EscrowedPaymentListItem from 'src/escrow/EscrowedPaymentListItem'
import { getReclaimableEscrowPayments } from 'src/escrow/saga'
import { updatePaymentRequestStatus } from 'src/firebase/actions'
import i18n, { Namespaces, withTranslation } from 'src/i18n'
import { fetchPhoneAddresses } from 'src/identity/actions'
import {
  NotificationList,
  titleWithBalanceNavigationOptions,
  useBalanceInNavigationParam,
} from 'src/notifications/NotificationList'
import { NumberToRecipient } from 'src/recipients/recipient'
import { recipientCacheSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'

interface StateProps {
  dollarBalance: string | null
  sentEscrowedPayments: EscrowedPayment[]
  recipientCache: NumberToRecipient
}

interface DispatchProps {
  updatePaymentRequestStatus: typeof updatePaymentRequestStatus
  fetchPhoneAddresses: typeof fetchPhoneAddresses
}

const mapStateToProps = (state: RootState): StateProps => ({
  dollarBalance: state.stableToken.balance,
  sentEscrowedPayments: getReclaimableEscrowPayments(state.escrow.sentEscrowedPayments),
  recipientCache: recipientCacheSelector(state),
})

type Props = NavigationInjectedProps & WithTranslation & StateProps & DispatchProps

export const listItemRenderer = (payment: EscrowedPayment, key: number | undefined = undefined) => {
  return (
    <View key={key}>
      <EscrowedPaymentListItem payment={payment} />
    </View>
  )
}

const EscrowedPaymentListScreen = (props: Props) => {
  const { dollarBalance, navigation } = props
  useBalanceInNavigationParam(dollarBalance, navigation)
  return (
    <NotificationList
      items={props.sentEscrowedPayments}
      listItemRenderer={listItemRenderer}
      dollarBalance={props.dollarBalance}
    />
  )
}

EscrowedPaymentListScreen.navigationOptions = titleWithBalanceNavigationOptions(
  i18n.t('walletFlow5:escrowedPaymentReminder')
)

export default connect<StateProps, DispatchProps, {}, RootState>(mapStateToProps, {
  updatePaymentRequestStatus,
  fetchPhoneAddresses,
})(withTranslation(Namespaces.global)(EscrowedPaymentListScreen))
