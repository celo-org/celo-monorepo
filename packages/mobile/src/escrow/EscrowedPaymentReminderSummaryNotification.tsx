import variables from '@celo/react-components/styles/variables'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Image, StyleSheet } from 'react-native'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { EscrowedPayment } from 'src/escrow/actions'
import EscrowedPaymentLineItem from 'src/escrow/EscrowedPaymentLineItem'
import { listItemRenderer } from 'src/escrow/EscrowedPaymentListScreen'
import { Namespaces } from 'src/i18n'
import { inviteFriendsIcon } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Stacks } from 'src/navigator/Screens'
import SummaryNotification from 'src/notifications/SummaryNotification'

interface OwnProps {
  payments: EscrowedPayment[]
}

type Props = OwnProps & WithNamespaces

export class EscrowedPaymentReminderSummaryNotification extends React.Component<Props> {
  onReview = () => {
    CeloAnalytics.track(CustomEventNames.escrowed_payment_review)
    navigate(Stacks.EscrowStack)
  }

  itemRenderer = (item: EscrowedPayment) => {
    return <EscrowedPaymentLineItem payment={item} key={item.paymentID} />
  }

  render() {
    const { payments, t } = this.props
    return payments.length === 1 ? (
      listItemRenderer(payments[0])
    ) : (
      <SummaryNotification<EscrowedPayment>
        items={payments}
        title={t('escrowedPaymentReminder')}
        icon={<Image source={inviteFriendsIcon} style={styles.image} resizeMode="contain" />}
        onReview={this.onReview}
        itemRenderer={this.itemRenderer}
      />
    )
  }
}

const styles = StyleSheet.create({
  body: {
    marginTop: 5,
    flexDirection: 'row',
  },
  image: {
    width: 30,
    height: 30,
  },
  requests: {
    flex: 1,
  },
  counter: {
    paddingLeft: variables.contentPadding,
    justifyContent: 'flex-end',
  },
})

export default withNamespaces(Namespaces.walletFlow5)(EscrowedPaymentReminderSummaryNotification)
