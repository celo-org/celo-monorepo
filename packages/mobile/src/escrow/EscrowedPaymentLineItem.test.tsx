import * as React from 'react'
import 'react-native'
import * as renderer from 'react-test-renderer'
import { escrowPaymentDouble } from 'src/escrow/__mocks__'
import EscrowedPaymentLineItem from 'src/escrow/EscrowedPaymentLineItem'

const mockedPayment = escrowPaymentDouble({})

it('renders correctly', () => {
  const tree = renderer.create(<EscrowedPaymentLineItem payment={mockedPayment} />)

  expect(tree).toMatchSnapshot()
})
