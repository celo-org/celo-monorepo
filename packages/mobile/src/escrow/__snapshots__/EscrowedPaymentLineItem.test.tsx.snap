// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  ellipsizeMode="middle"
  numberOfLines={1}
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#81868B",
          "fontFamily": "Hind-Regular",
          "fontSize": 14,
          "lineHeight": 20,
        },
      ]
    }
  >
    +491522345678
     - 
    test message
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#81868B",
          "fontFamily": "Hind-Regular",
          "fontSize": 14,
          "lineHeight": 20,
        },
        Object {
          "fontFamily": "Hind-SemiBold",
        },
      ]
    }
  >
     
    $
    7
  </Text>
</Text>
`;
